pkg/controllerutils/patch.go

```diff
@@ -74,33 +70,25 @@ func getAndCreateOrPatch(ctx context.Context, c client.Client, obj client.Object
 }
 
-// CreateOrMergePatch creates or patches (using a merge patch) the given object in the Kubernetes cluster.
-// The object's desired state is only reconciled with the existing state inside the passed in callback MutateFn,
-// however, the object is not read from the client. This means the object should already be filled with the
-// last-known state if operating on more complex structures (e.g. if the patch is supposed to remove an optional field
-// or section).
+// CreateOrGetAndMergePatch creates or gets and patches (using a merge patch) the given object in the Kubernetes cluster.
 //
 // The MutateFn is called regardless of creating or patching an object.
 //
 // It returns the executed operation and an error.
-func CreateOrMergePatch(ctx context.Context, c client.Writer, obj client.Object, f controllerutil.MutateFn) (controllerutil.OperationResult, error) {
-	return createOrPatch(ctx, c, obj, func(obj client.Object) client.Patch { return client.MergeFrom(obj) }, f)
+func CreateOrGetAndMergePatch(ctx context.Context, c client.Client, obj client.Object, f controllerutil.MutateFn) (controllerutil.OperationResult, error) {
+	return createOrGetAndPatch(ctx, c, obj, func(obj client.Object) client.Patch { return client.MergeFrom(obj) }, f)
 }
 
-// CreateOrStrategicMergePatch creates or patches (using a strategic merge patch) the given object in the Kubernetes cluster.
-// The object's desired state is only reconciled with the existing state inside the passed in callback MutateFn,
-// however, the object is not read from the client. This means the object should already be filled with the
-// last-known state if operating on more complex structures (e.g. if the patch is supposed to remove an optional field
-// or section).
+// CreateOrGetAndStrategicMergePatch creates or gets and patches (using a strategic merge patch) the given object in the Kubernetes cluster.
 //
 // The MutateFn is called regardless of creating or patching an object.
 //
 // It returns the executed operation and an error.
-func CreateOrStrategicMergePatch(ctx context.Context, c client.Writer, obj client.Object, f controllerutil.MutateFn) (controllerutil.OperationResult, error) {
-	return createOrPatch(ctx, c, obj, func(obj client.Object) client.Patch { return client.StrategicMergeFrom(obj) }, f)
+func CreateOrGetAndStrategicMergePatch(ctx context.Context, c client.Client, obj client.Object, f controllerutil.MutateFn) (controllerutil.OperationResult, error) {
+	return createOrGetAndPatch(ctx, c, obj, func(obj client.Object) client.Patch { return client.StrategicMergeFrom(obj) }, f)
 }
 
-func createOrPatch(ctx context.Context, c client.Writer, obj client.Object, patchFunc func(client.Object) client.Patch, f controllerutil.MutateFn) (controllerutil.OperationResult, error) {
+func createOrGetAndPatch(ctx context.Context, c client.Client, obj client.Object, patchFunc func(client.Object) client.Patch, f controllerutil.MutateFn) (controllerutil.OperationResult, error) {
```

Comment:
As discussed, we need to 
- calculate the patch based on the object after we read it from the API server / cache (i.e. after `Get` but before `f()`)
- reset the object to an empty object (obtain it from `c.Scheme().New`) when we received an `AlreadyExistsError` to unset optional fields that were set in `f()`. If we don't unset optional fields added in `f()`, `Get` will not unset them for us and thus we won't set them via the patch. We need something like this: https://github.com/gardener/gardener/blob/893f4042fa3e4a1814ab04b4dd79e9a09e1f914a/pkg/extensions/customresources.go#L520-L543

Reply:
Thanks, done!